src/enums.ts

```diff
@@ -72,12 +81,22 @@ export function generateEnumFromJson(ctx: Context, fullName: string, enumDesc: E
   const chunks: Code[] = [];
 
   const functionName = camelCase(fullName) + 'FromJSON';
-  chunks.push(code`export function ${def(functionName)}(object: any): ${fullName} {`);
+  const returnType = options.enumUnspecifiedAsUndefined ? `${fullName} | undefined` : fullName;
+  chunks.push(code`export function ${def(functionName)}(object: any): ${returnType} {`);
   chunks.push(code`switch (object) {`);
 
   for (const valueDesc of enumDesc.value) {
     chunks.push(code`
-      case ${valueDesc.number}:
+      case ${valueDesc.number}:`);
+
+    if (options.enumUnspecifiedAsUndefined && valueDesc.number === 0) {
+      chunks.push(code`
+      case undefined:
+        return undefined`);
+      continue;
```

Comment:
Shouldn't we leave this `continue` out? Because if some other system (that is not ts-proto) sends us a enum with the 0 value, but as the string "FOO_UNKNOWN", we'd want to still treat it as undefined.

I.e. something like:

```
  switch (object) {
    case 0:
    case 'UNKNOWN':
    case undefined:
      return undefined;
    case 2:
    case 'ON':
```

Granted, if the other system is itself ts-proto, or using the proto3 rules to not include the enum value if its the 0th value anyway, this extra "case 'UNKNOWN'" would be strictly required, but it seems like a good idea to include just in case?

Also, do we need to include "case undefined"? I don't think its part of the proto3 JSON spec to send `state: undefined` on the wire.

I see that you've changed `toJSON` to also return `undefined`:

```

export function stateEnumToJSON(object: StateEnum | undefined): string | undefined {
  switch (object) {
    case undefined:
      return undefined;
    case StateEnum.ON:
      return 'ON';
    case StateEnum.OFF:
      return 'OFF';
    case StateEnum.UNRECOGNIZED:
    default:
      return 'UNRECOGNIZED';
  }
}
```

But I don't think we should do this; i.e. the `enumUnspecificAsUndefined` should only control how ts-proto messages are represented internally and not change how messages are encoded/decoded from JSON on the wire.

Reply:
You're right, I have not thought about that👍

Reply:
> Also, do we need to include "case undefined"? I don't think its part of the proto3 JSON spec to send state: undefined on the wire.

Just in case for some other use case, I think we should not send 'UNRECOGNIZED' or throw an error (when unrecognizedEnum=false) when value is undefined?

Reply:
> I see that you've changed toJSON to also return undefined:

Because I'd removed the default(unspecified) value from enum definition, if `toJSON` output `XXX_UNSPECIFIED`, that is inconsistent behavior, unless don't remove the default value from enum?
But also strange, the default value already decoded as `undefined`, why keep the default value in enum?